expectedExceptions mismatch because of wrapping (regression 7.0.0 to 7.1.0)

### TestNG Version
7.1.0

### Expected behavior
When the method under test throws an IllegalArgumentException, a test with `expectedExceptions = IllegalArgumentException.class` passes.

### Actual behavior
The test fails with:
```
FAILED: testConvertWrongNumberOfBytesToInts
org.testng.TestException: 
Expected exception of type class java.lang.IllegalArgumentException but got org.testng.internal.InvokeMethodRunnable$TestNGRuntimeException: java.lang.IllegalArgumentException: Number of input bytes must be a multiple of 4.
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:816)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:146)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.testng.internal.InvokeMethodRunnable$TestNGRuntimeException: java.lang.IllegalArgumentException: Number of input bytes must be a multiple of 4.
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:53)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:71)
	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:35)
	at org.testng.internal.MethodInvocationHelper.invokeWithTimeoutWithNoExecutor(MethodInvocationHelper.java:314)
	at org.testng.internal.MethodInvocationHelper.invokeWithTimeout(MethodInvocationHelper.java:280)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:603)
	... 9 more
Caused by: java.lang.IllegalArgumentException: Number of input bytes must be a multiple of 4.
	at io.github.pr0methean.betterrandom.util.BinaryUtils.convertBytesToInts(BinaryUtils.java:104)
	at io.github.pr0methean.betterrandom.util.BinaryUtilsTest.testConvertWrongNumberOfBytesToInts(BinaryUtilsTest.java:96)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:134)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
```

### Is the issue reproductible on runner?

- [ ] Shell
- [x] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [ ] NetBeans


### Test case sample
Test:
https://github.com/Pr0methean/BetterRandom/blob/6df10c5b6e8ec24dea4275bf2b487dd9aecdc369/betterrandom/src/test/java/io/github/pr0methean/betterrandom/util/BinaryUtilsTest.java#L93
Code under test throws the exception here:
https://github.com/Pr0methean/BetterRandom/blob/ca5deaff0497db33d7769bdf697af9852389f151/betterrandom/src/main/java/io/github/pr0methean/betterrandom/util/BinaryUtils.java#L104


(This version of the test passes on TestNG 7.0.0.)

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,5 @@
 Current
+Fixed: GITHUB-2235: expectedExceptions mismatch because of wrapping (Krishnan Mahadevan)
 Fixed: GITHUB-2220: ITestListener's methods get called multiple times for one test, when @Listeners annotation is used in multiple test classes (Krishnan Mahadevan)
 Fixed: GITHUB-2211: assertEquals for Map sometimes does not use provided message (Krishnan Mahadevan)
 Fixed: GITHUB-1632: throwing SkipException sets iTestResult status to Failure instead of Skip (Krishnan Mahadevan)

--- a/src/main/java/org/testng/internal/MethodInvocationHelper.java
+++ b/src/main/java/org/testng/internal/MethodInvocationHelper.java
@@ -12,6 +12,7 @@ import org.testng.ITestContext;
 import org.testng.ITestNGMethod;
 import org.testng.ITestResult;
 import org.testng.TestNGException;
+import org.testng.internal.InvokeMethodRunnable.TestNGRuntimeException;
 import org.testng.internal.annotations.IAnnotationFinder;
 import org.testng.internal.collections.ArrayIterator;
 import org.testng.internal.collections.OneToTwoDimArrayIterator;
@@ -328,7 +329,11 @@ public class MethodInvocationHelper {
       }
     } catch (Exception ex) {
       if (notTimedout) {
-        testResult.setThrowable(ex.getCause());
+        Throwable e = ex.getCause();
+        if (e instanceof TestNGRuntimeException) {
+          e = e.getCause();
+        }
+        testResult.setThrowable(e);
       } else {
         ThreadTimeoutException exception =
             new ThreadTimeoutException(

--- a/src/test/java/test/expectedexceptions/issue2235/ExampleTestCase.java
+++ b/src/test/java/test/expectedexceptions/issue2235/ExampleTestCase.java
@@ -1,0 +1,11 @@
+package test.expectedexceptions.issue2235;
+
+import org.testng.annotations.Test;
+
+public class ExampleTestCase {
+
+  @Test(timeOut = 1000, expectedExceptions = IllegalArgumentException.class)
+  public void testMethod() {
+    throw new IllegalArgumentException("foo");
+  }
+}

--- a/src/test/java/test/expectedexceptions/issue2235/IssueTest.java
+++ b/src/test/java/test/expectedexceptions/issue2235/IssueTest.java
@@ -1,0 +1,25 @@
+package test.expectedexceptions.issue2235;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.testng.TestListenerAdapter;
+import org.testng.TestNG;
+import org.testng.annotations.Test;
+import org.testng.xml.XmlSuite;
+import org.testng.xml.XmlSuite.ParallelMode;
+import test.SimpleBaseTest;
+
+public class IssueTest extends SimpleBaseTest {
+
+  @Test(description = "GITHUB-2235")
+  public void testExpectedExceptions() {
+    XmlSuite xmlSuite = createXmlSuite("main_suite", "main_test", ExampleTestCase.class);
+    xmlSuite.setParallel(ParallelMode.METHODS);
+    TestNG testng = create(xmlSuite);
+    TestListenerAdapter listener = new TestListenerAdapter();
+    testng.addListener(listener);
+    testng.run();
+    assertThat(listener.getPassedTests()).isNotEmpty();
+    assertThat(listener.getFailedTests()).isEmpty();
+  }
+}

--- a/src/test/resources/testng.xml
+++ b/src/test/resources/testng.xml
@@ -61,6 +61,7 @@
       <class name="test.CtorCalledOnce" />
       <class name="test.expectedexceptions.ExpectedExceptionsTest" />
       <class name="test.expectedexceptions.issue2074.IssueTest"/>
+      <class name="test.expectedexceptions.issue2235.IssueTest"/>
       <class name="test.access.PrivateAccessConfigurationMethods" />
       <class name="test.expectedexceptions.WrappedExpectedExceptionTest" />
       <class name="test.parameters.OptionalParameterTest"/>

